cpp/example/simple_kv_store.cc

```diff
@@ -0,0 +1,210 @@
+  std::unique_lock<std::mutex> lock(mtx_);
+  return data_;
+}
+
+void BackupServer::SyncData(const std::string &key, const std::string &val) {
+  std::unique_lock<std::mutex> lock(mtx_);
+  data_[key] = val;
+}
+
+class MainServer {
+ public:
+  MainServer();
+
+  void Put(const std::string &key, const std::string &val);
+
+  std::pair<bool, std::string> Get(const std::string &key);
+
+  std::unordered_map<std::string, std::string> GetAllData();
+
+ private:
+  void HanldeFailover();
+
+  std::unordered_map<std::string, std::string> data_;
+  std::mutex mtx_;
+
+  ray::ActorHandle<BackupServer> backup_actor_;
+};
+
+MainServer::MainServer() {
+  if (ray::WasCurrentActorRestarted()) {
+    HanldeFailover();
+  }
+
+  backup_actor_ = *ray::GetActor<BackupServer>(BACKUP_SERVER_NAME);
+  RAYLOG(INFO) << "MainServer created";
+}
+
+void MainServer::Put(const std::string &key, const std::string &val) {
+  // SyncData before put data.
+  auto r = backup_actor_.Task(&BackupServer::SyncData).Remote(key, val);
+  std::vector<ray::ObjectRef<void>> objects{r};
+  auto result = ray::Wait(objects, 1, 2000);
+  if (result.ready.empty()) {
+    RAYLOG(WARNING) << "MainServer SyncData failed.";
+  }
+
+  std::unique_lock<std::mutex> lock(mtx_);
+  data_[key] = val;
+}
+
+std::pair<bool, std::string> MainServer::Get(const std::string &key) {
+  std::unique_lock<std::mutex> lock(mtx_);
+  return common::Get(key, data_);
+}
+
+std::unordered_map<std::string, std::string> MainServer::GetAllData() {
+  std::unique_lock<std::mutex> lock(mtx_);
+  return data_;
+}
+
+void BackupServer::HanldeFailover() {
+  // Get all data from MainServer.
+  auto dest_actor = *ray::GetActor<MainServer>(MAIN_SERVER_NAME);
+  data_ = *dest_actor.Task(&MainServer::GetAllData).Remote().Get();
+  RAYLOG(INFO) << "BackupServer get all data from MainServer";
+}
+
+void MainServer::HanldeFailover() {
+  // Get all data from BackupServer.
+  backup_actor_ = *ray::GetActor<BackupServer>(BACKUP_SERVER_NAME);
+  data_ = *backup_actor_.Task(&BackupServer::GetAllData).Remote().Get();
+  RAYLOG(INFO) << "MainServer get all data from BackupServer";
+}
+
+static MainServer *CreateMainServer() { return new MainServer(); }
+
+static BackupServer *CreateBackupServer() { return new BackupServer(); }
+
+RAY_REMOTE(CreateMainServer, &MainServer::GetAllData, &MainServer::Get, &MainServer::Put);
+
+RAY_REMOTE(CreateBackupServer, &BackupServer::GetAllData, &BackupServer::SyncData);
+
+ray::PlacementGroup CreateSimplePlacementGroup(const std::string &name) {
+  std::vector<std::unordered_map<std::string, double>> bundles{
+      {{"CPU", 1.0}, {"memory", 1024.0}}, {{"CPU", 1.0}, {"memory", 1024.0}}};
+
+  ray::PlacementGroupCreationOptions options{false, name, bundles,
+                                             ray::PlacementStrategy::SPREAD};
+  return ray::CreatePlacementGroup(options);
+}
+
+void StartServer() {
+  auto placement_group = CreateSimplePlacementGroup("my_placement_group");
```

Comment:
more comments please

Reply:
Something like
"Create the main server actor and backup server actor. Assuming that each of them needs 1 cpu core and 1 GB memory. And we use Ray's placement group to make sure that the 2 actors are created nodes if possible."